lightning/src/util/ser_macros.rs

```diff
@@ -7,82 +7,142 @@
+			BigSize(field_len as u64).write(&mut $len).expect("No in-memory data may fail to serialize");
+			$len.0 += field_len;
+		}
+	};
+}
+
 macro_rules! get_varint_length_prefixed_tlv_length {
-	({$(($type: expr, $field: expr)),*}, {$(($optional_type: expr, $optional_field: expr)),* $(,)*}) => { {
+	({$(($type: expr, $field: expr, $fieldty: ident)),*}) => { {
 		use util::ser::LengthCalculatingWriter;
 		#[allow(unused_mut)]
 		let mut len = LengthCalculatingWriter(0);
 		{
 			$(
-				BigSize($type).write(&mut len).expect("No in-memory data may fail to serialize");
-				let field_len = $field.serialized_length();
-				BigSize(field_len as u64).write(&mut len).expect("No in-memory data may fail to serialize");
-				len.0 += field_len;
-			)*
-			$(
-				if let Some(ref field) = $optional_field {
-					BigSize($optional_type).write(&mut len).expect("No in-memory data may fail to serialize");
-					let field_len = field.serialized_length();
-					BigSize(field_len as u64).write(&mut len).expect("No in-memory data may fail to serialize");
-					len.0 += field_len;
-				}
+				get_a_varint_length_prefixed_tlv_length!(len, $type, $field, $fieldty);
 			)*
 		}
 		len.0
 	} }
 }
 
 macro_rules! encode_varint_length_prefixed_tlv {
-	($stream: expr, {$(($type: expr, $field: expr)),*}, {$(($optional_type: expr, $optional_field: expr)),*}) => { {
+	($stream: expr, {$(($type: expr, $field: expr, $fieldty: ident)),*}) => { {
 		use util::ser::BigSize;
-		let len = get_varint_length_prefixed_tlv_length!({ $(($type, $field)),* }, { $(($optional_type, $optional_field)),* });
+		let len = get_varint_length_prefixed_tlv_length!({ $(($type, $field, $fieldty)),* });
 		BigSize(len as u64).write($stream)?;
-		encode_tlv!($stream, { $(($type, $field)),* }, { $(($optional_type, $optional_field)),* });
+		encode_tlv!($stream, { $(($type, $field, $fieldty)),* });
 	} }
 }
 
+macro_rules! check_tlv_order {
+	($last_seen_type: expr, $typ: expr, $type: expr, explicit) => {{
+		#[allow(unused_comparisons)] // Note that $type may be 0 making the second comparison always true
+		let invalid_order = ($last_seen_type.is_none() || $last_seen_type.unwrap() < $type) && $typ.0 > $type;
+		if invalid_order {
+			Err(DecodeError::InvalidValue)?
+		}
+	}};
+	($last_seen_type: expr, $typ: expr, $type: expr, option) => {{
+		// no-op
+	}};
+	($last_seen_type: expr, $typ: expr, $type: expr, vec_type) => {{
+		// no-op
+	}};
+}
+
+macro_rules! check_missing_tlv {
+	($last_seen_type: expr, $type: expr, explicit) => {{
+		#[allow(unused_comparisons)] // Note that $type may be 0 making the second comparison always true
+		let missing_req_type = $last_seen_type.is_none() || $last_seen_type.unwrap() < $type;
+		if missing_req_type {
+			Err(DecodeError::InvalidValue)?
+		}
+	}};
+	($last_seen_type: expr, $type: expr, vec_type) => {{
+		// no-op
+	}};
+	($last_seen_type: expr, $type: expr, option) => {{
+		// no-op
+	}};
+}
+
+macro_rules! decode_a_tlv {
```

Comment:
Consider renaming `decode_tlv` to `decode_tlv_stream` and name this `decode_tlv`. Similarly elsewhere when the only difference in naming is `_a_`.

Reply:
Good one